openssl/src/x509/extension.rs

```diff
@@ -428,6 +428,145 @@ impl AuthorityKeyIdentifier {
     }
 }
 
+#[cfg(ossl110)]
+pub struct SbgpAsIdentifier {
+    critical: bool,
+    asn: Vec<(u32, u32)>,
+}
+
+#[cfg(ossl110)]
+impl Default for SbgpAsIdentifier {
+    fn default() -> SbgpAsIdentifier {
+        SbgpAsIdentifier::new()
+    }
+}
+
+#[cfg(ossl110)]
+impl SbgpAsIdentifier {
+    /// Construct a new `SbgpAsIdentifier` extension.
+    pub fn new() -> SbgpAsIdentifier {
+        SbgpAsIdentifier {
+            critical: false,
+            asn: Vec::new(),
+        }
+    }
+
+    /// Sets the `critical` flag to `true`. The extension will be critical.
+    pub fn critical(&mut self) -> &mut SbgpAsIdentifier {
+        self.critical = true;
+        self
+    }
+
+    /// Adds an AS number.
+    pub fn add_asn(&mut self, asn: u32) -> &mut SbgpAsIdentifier {
+        self.asn.push((asn, asn));
+        self
+    }
+
+    /// Adds a range of AS numbers.
+    pub fn add_asn_range(&mut self, asn_min: u32, asn_max: u32) -> &mut SbgpAsIdentifier {
+        self.asn.push((asn_min, asn_max));
+        self
+    }
+
+    /// Return a `SbgpAsIdentifier` extension as an `X509Extension`.
+    #[allow(deprecated)]
+    pub fn build(&self, ctx: &X509v3Context<'_>) -> Result<X509Extension, ErrorStack> {
+        let mut value = String::new();
+        let mut first = true;
+        append(&mut value, &mut first, self.critical, "critical");
+        for (asn_min, asn_max) in &self.asn {
+            let asn_format = if asn_min == asn_max {
+                format!("AS:{}", asn_min)
+            } else {
+                format!("AS:{}-{}", asn_min, asn_max)
+            };
+            append(&mut value, &mut first, true, &asn_format);
+        }
+        X509Extension::new_nid(None, Some(ctx), Nid::SBGP_AUTONOMOUSSYSNUM, &value)
```

Comment:
We need to move away from the string APIs for x509 extensions, and instead directly work with the relevant OpenSSL types.